impl-tinkerpop/src/main/java/org/hawkular/inventory/impl/tinkerpop/FilterVisitor.java

```diff
@@ -109,4 +112,100 @@ public void visit(HawkularPipeline<?, ?> query, With.Types types) {
 
         query.or(typeChecks);
     }
+
+    public void visit(HawkularPipeline<?, ?> query, RelationWith.Ids ids) {
+        if (ids.getIds().length == 1) {
+            query.has("id", ids.getIds()[0]);
+            return;
+        }
+
+        Pipe[] idChecks = new Pipe[ids.getIds().length];
+
+        Arrays.setAll(idChecks, i ->
+                new PropertyFilterPipe<Element, String>("id", Compare.EQUAL, ids.getIds()[i]));
+
+        query.or(idChecks);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void visit(HawkularPipeline<?, ?> query, RelationWith.Properties properties) {
+        if (properties.getValues().length == 1) {
+            query.has(properties.getProperty(), properties.getValues()[0]);
+            return;
+        }
+
+        Pipe[] idChecks = new Pipe[properties.getValues().length];
+
+        Arrays.setAll(idChecks, i ->
+                new PropertyFilterPipe<Element, String>(properties.getProperty(), Compare.EQUAL, properties
+                        .getValues()[i]));
+
+        query.or(idChecks);
+    }
+
+    public void visit(HawkularPipeline<?, ?> query, RelationWith.SourceOfType types) {
+        visit(query, types, true);
+    }
+
+    public void visit(HawkularPipeline<?, ?> query, RelationWith.TargetOfType types) {
+        visit(query, types, false);
+    }
+
+    public void visit(HawkularPipeline<?, ?> query, RelationWith.SourceOrTargetOfType types) {
+        visit(query, types, null);
+    }
+
+    private void visit(HawkularPipeline<?, ?> query, RelationWith.SourceOrTargetOfType types, Boolean source) {
+        // look ahead if the type of the incidence vertex is of the desired type(s)
+        String label = UUID.randomUUID().toString();
+        HawkularPipeline<?, ?> q1 = query.as(label);
+        HawkularPipeline<?, ?> q2;
+        if (source == null) {
+            q2 = q1.bothV();
+        } else if (source) {
+            q2 = q1.outV();
+        } else {
+            q2 = q1.inV();
+        }
+        if (types.getTypes().length == 1) {
+            Constants.Type type = Constants.Type.of(types.getTypes()[0]);
+            q2.has(Constants.Property.type.name(), type.name()).back(label);
+            return;
+        }
+
+        Pipe[] typeChecks = new Pipe[types.getTypes().length];
+        Arrays.setAll(typeChecks, i -> {
+            Constants.Type type = Constants.Type.of(types.getTypes()[i]);
+            return new PropertyFilterPipe<Element, String>(Constants.Property.type.name(), Compare.EQUAL, type.name());
+        });
+
+        q2.or(typeChecks).back(label);
+    }
+
+    public void visit(HawkularPipeline<?, ?> query, RelationshipBrowser.JumpInOutFilter filter) {
+        final boolean jumpFromEdge = filter.isFromEdge();
+        switch (filter.getDirection()) {
+            case incoming:
+                if (jumpFromEdge) {
+                    query.outV();
```

Comment:
Why is an incoming direction of a relationship modelled as jumping to the "out" vertex of an edge?